src/runtime-rs/src/shim/src/task_service.rs

```diff
@@ -0,0 +1,848 @@
+// Copyright (c) 2021 Alibaba Cloud
+// Copyright (c) 2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use ::log::{debug, error, info, trace, warn};
+
+use ttrpc::error::get_rpc_status;
+use ttrpc::{Code, Result, TtrpcContext};
+use virtcontainers::Sandbox;
+
+use shim_proto::task::{ProcessInfo, Status};
+use shim_proto::{empty::Empty, metrics, shim, shim_ttrpc::Task};
+use std::convert::{TryFrom, TryInto};
+
+use protobuf::Message;
+
+use agent::{BlkioStatsEntry, StatsContainerResponse};
+use nix::sys::signal::Signal;
+use std::sync::mpsc::Sender;
+use std::sync::{Arc, Mutex};
+use virtcontainers::container::State;
+use virtcontainers::oci::ContainerType;
+
+use crate::run::{ServerAction, ServerMessage};
+
+#[allow(dead_code)]
+pub(crate) struct TaskService {
+    pub id: String,
+    pub pid: u32,
+    pub sandbox: Arc<Mutex<Sandbox>>,
+    pub shutdown_sender: Mutex<Sender<ServerMessage>>,
+}
+
+impl Task for TaskService {
+    fn state(&self, _ctx: &TtrpcContext, req: shim::StateRequest) -> Result<shim::StateResponse> {
+        debug!("====> state: {:?}", req);
+        let rsp = self.do_state(req);
+        debug!("<==== state: {:?}", rsp);
+        rsp
+    }
+
+    fn create(
+        &self,
+        _ctx: &TtrpcContext,
+        req: shim::CreateTaskRequest,
+    ) -> Result<shim::CreateTaskResponse> {
+        debug!("====> create: {:?}", req);
+        let rsp = self.do_create(req);
+        debug!("<==== create: {:?}", rsp);
+        rsp
```

Comment:
tbh, most of these functions are so similar, I think it might make sense to create a macro to generate them. I keep meaning to auto-generate more of https://github.com/kata-containers/kata-containers/blob/main/tools/agent-ctl/src/client.rs in a similar way.

Reply:
I am not sure about auto generation.  Although it would reduce similar codes,  using a macro also make it more difficult for to modify or debug a specific function.